### DISCUSSION TITLE
IR drop analysis reports negative values

### DISCUSSION THREAD

**Discussion by ishwars1618** [The-OpenROAD-Project/OpenROAD-flow-scripts#1789] [Q&A]
When using OpenROAD flow on certain circuits (usually large circuits, having around 10^4 cell instances), I found that the `analyze_power_grid` tool reports negative voltage values at each instance, even though I had set the VDD to 0.77V. 

I have attached a package with the test case below for the `final_report` step.

Can you please let me know how this can be resolved?

Thanks,
Ishwar S.

Please download the test case package from here: [https://drive.google.com/file/d/1BpxVOgMD08aVjMaJrdFYml4kR0VSF-jj/view?usp=drive_link](https://drive.google.com/file/d/1BpxVOgMD08aVjMaJrdFYml4kR0VSF-jj/view?usp=drive_link)

**Comment by maliberty** [comment]
I don't see a negative value reported:
```
########## IR report #################
Corner: default
Worstcase voltage: 1.69e-01 V
Average IR drop  : 4.60e-01 V
Worstcase IR drop: 6.01e-01 V
######################################
########## IR report #################
Corner: default
Worstcase voltage: 5.93e-01 V
Average IR drop  : 4.60e-01 V
Worstcase IR drop: 5.93e-01 V
######################################
```

How are you observing this negative value?

**Reply by ishwars1618** [reply]
I see negative values in `results/asap7/tc1_1e4/base/6_instance_voltage_pwr.txt`, which is produced by `scripts/info-beta.tcl` in line 76. The vsrc file used in `analyze_power_grid` can be seen [here](https://drive.google.com/file/d/13uKXRJNx1WsTugp4C7Lf7Ie7oOWOJEoe/view?usp=drive_link). This contains vsrc nodes with 0.7V instead of 0.77V. 
I changed both vsrcs to 0.77V and reran the 6th step, but this still results in negative voltage values, which can be seen in this [file](https://drive.google.com/file/d/1XWXPztxlRtef-e0Ys-Ic0_mKj7b30qYN/view?usp=drive_link).

**Comment by maliberty** [comment]
Without the vsrc file I don't see the negative values (the test case was not setup this way as originally packaged).  With it I do.  The file is incorrect as it only has three values per line not four as described in https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/src/psm/doc/Vsrc_description.md

I'll add some error checking to the file reader.

**Reply by maliberty** [reply]
I see you can leave off the fourth value but in any case the third value is the bump size not the voltage.  For example with
```
10,10,0.7
40,40,0.7
```
I don't see any negative values

**Comment by b224hisl** [comment]
For IR-related, I would like to ask what the variable `IR_DROP_LAYER` is used for. Is the IR drop different across various metal layers, and does it need to be distinguished? If so, it seems that this variable is not specified during power grid analysis （`analyze_power_grid`）. It appears that this variable is only used in the TCL script for saving images. (`save_images.tcl`)

**Reply by maliberty** [reply]
It is just the layer on which we save an image of the IR drop during the flow.  The calculation is done on all layers.  Typically the drop is most interesting on metal1 but the name varies by PDK.

**Reply by b224hisl** [reply]
Fine, thank you very much! I have another question. If I don't define the `voltage_source_file` by `—src` when executing `analyze_power_grid,` what are the default locations of the C4 bumps?

**Reply by maliberty** [reply]
The defaults are mention in https://openroad.readthedocs.io/en/latest/main/src/psm/README.html#set-pdnsim-power-source-settings

 The default bump pitch is 140um.
 The default bump size is 70um.
 The default bump interval is 3.

**Reply by b224hisl** [reply]
You replied so quickly, thank you very much.
